x/auth/tx/builder.go

```diff
@@ -0,0 +1,317 @@
+	if len(t.authInfoBz) == 0 {
+		// if authInfoBz is empty, then marshal the body. authInfoBz will generally
+		// be set to nil whenever SetAuthInfo is called so the result of calling
+		// this method should always return the correct bytes. Note that after
+		// decoding authInfoBz is derived from TxRaw so that it matches what was
+		// transmitted over the wire
+		t.authInfoBz = t.marshaler.MustMarshalBinaryBare(t.tx.AuthInfo)
+	}
+	return t.authInfoBz
+}
+
+func (t *builder) GetSigners() []sdk.AccAddress {
+	var signers []sdk.AccAddress
+	seen := map[string]bool{}
+
+	for _, msg := range t.GetMsgs() {
+		for _, addr := range msg.GetSigners() {
+			if !seen[addr.String()] {
+				signers = append(signers, addr)
+				seen[addr.String()] = true
+			}
+		}
+	}
+
+	return signers
+}
+
+func (t *builder) GetPubKeys() []crypto.PubKey {
+	if t.pubKeys == nil {
+		signerInfos := t.tx.AuthInfo.SignerInfos
+		pubKeys := make([]crypto.PubKey, len(signerInfos))
+
+		for i, si := range signerInfos {
+			var err error
+			pk := si.PublicKey
+			if pk != nil {
+				pubKeys[i], err = t.pubkeyCodec.Decode(si.PublicKey)
+				if err != nil {
+					panic(err)
+				}
+			}
+		}
+
+		t.pubKeys = pubKeys
+	}
+
+	return t.pubKeys
+}
+
+func (t *builder) GetGas() uint64 {
+	return t.tx.AuthInfo.Fee.GasLimit
+}
+
+func (t *builder) GetFee() sdk.Coins {
+	return t.tx.AuthInfo.Fee.Amount
+}
+
+func (t *builder) FeePayer() sdk.AccAddress {
+	return t.GetSigners()[0]
+}
+
+func (t *builder) GetMemo() string {
+	return t.tx.Body.Memo
+}
+
+func (t *builder) GetSignatures() [][]byte {
+	return t.tx.Signatures
+}
+
+func (t *builder) GetSignaturesV2() ([]signing.SignatureV2, error) {
+	signerInfos := t.tx.AuthInfo.SignerInfos
+	sigs := t.tx.Signatures
+	pubKeys := t.GetPubKeys()
+	n := len(signerInfos)
+	res := make([]signing.SignatureV2, n)
+
+	for i, si := range signerInfos {
+		var err error
+		sigData, err := ModeInfoAndSigToSignatureData(si.ModeInfo, sigs[i])
+		if err != nil {
+			return nil, err
+		}
+		res[i] = signing.SignatureV2{
+			PubKey: pubKeys[i],
+			Data:   sigData,
+		}
+	}
+
+	return res, nil
+}
+
+func (t *builder) SetMsgs(msgs ...sdk.Msg) error {
+	anys := make([]*codectypes.Any, len(msgs))
+
+	for i, msg := range msgs {
+		var err error
+		anys[i], err = codectypes.NewAnyWithValue(msg)
+		if err != nil {
+			return err
+		}
+	}
+
+	t.tx.Body.Messages = anys
+
+	// set bodyBz to nil because the cached bodyBz no longer matches tx.Body
+	t.bodyBz = nil
+
+	return nil
+}
+
+func (t *builder) SetMemo(memo string) {
+	t.tx.Body.Memo = memo
+
+	// set bodyBz to nil because the cached bodyBz no longer matches tx.Body
+	t.bodyBz = nil
+}
+
+func (t *builder) SetGasLimit(limit uint64) {
+	if t.tx.AuthInfo.Fee == nil {
+		t.tx.AuthInfo.Fee = &tx.Fee{}
+	}
+
+	t.tx.AuthInfo.Fee.GasLimit = limit
+
+	// set authInfoBz to nil because the cached authInfoBz no longer matches tx.AuthInfo
+	t.authInfoBz = nil
+}
+
+func (t *builder) SetFeeAmount(coins sdk.Coins) {
```

Comment:
@aaronc wouldn't it be better to have SetFee rather than just feeAmount

Reply:
`SetFeeAmount` was requested in a previous PR for `TxBuilder`. I think it's okay but either way is out of the scope of this PR.